stl/inc/algorithm

```diff
@@ -961,6 +941,322 @@ namespace ranges {
+                if (_First1 == _End1) { // first range is a (possibly proper?) prefix of the second
+                    return _First2 == _End2;
+                } else if (_First2 == _End2) { // second range is a proper prefix of the first
+                    return false;
+                }
+
+                if (!_STD invoke(_Pred, _STD invoke(_Proj1, *_First1), _STD invoke(_Proj2, *_First2))) { // mismatch
+                    break;
+                }
+            }
+            // If we get here, initial elements do not match.
+
+            // Determine final iterator values and validate lengths
+            auto _Last1 = _First1;
+            auto _Last2 = _First2;
+            for (;;) {
+                ++_Last1;
+                ++_Last2;
+                if (_Last1 == _End1) {
+                    if (_Last2 == _End2) {
+                        break; // equal lengths
+                    }
+
+                    return false; // different lengths; not a permutation
+                } else if (_Last2 == _End2) {
+                    return false; // Ditto different lengths
+                }
+            }
+            // if we get here, initial elements do not match and ranges have equal lengths.
+
+            if constexpr (bidirectional_iterator<_It1> && bidirectional_iterator<_It2>) {
+                for (;;) { // Trim matching suffixes
+                    if (--_Last1 == _First1) {
+                        return false; // initial elements did not match when trimming prefixes.
+                    }
+                    if (--_Last2 == _First2) {
+                        return false; // Ditto
+                    }
+                    if (!_STD invoke(_Pred, _STD invoke(_Proj1, *_Last1), _STD invoke(_Proj2, *_Last2))) { // mismatch
+                        ++_Last1;
+                        ++_Last2;
+                        break;
+                    }
+                }
+                // If we get here, initial elements do not match, final elements do not match, and ranges have length
+                // at least 2.
+            }
+
+            return _Match_counts(
+                _STD move(_First1), _STD move(_Last1), _STD move(_First2), _STD move(_Last2), _Pred, _Proj1, _Proj2);
+        }
+
+        template <class _It1, class _Se1, class _It2, class _Se2, class _Pr, class _Pj1, class _Pj2>
+        _NODISCARD static constexpr bool _Match_counts(const _It1 _First1, const _Se1 _Last1, const _It2 _First2,
+            const _Se2 _Last2, _Pr _Pred, _Pj1 _Proj1, _Pj2 _Proj2) {
+            for (auto _Mid1 = _First1; _Mid1 != _Last1; ++_Mid1) {
+                bool _Found = false;
+                auto _Meow  = _First1;
+                for (; _Meow != _Mid1; ++_Meow) {
+                    if (_STD invoke(_Pred, _STD invoke(_Proj1, *_Mid1), _STD invoke(_Proj1, *_Meow))) {
+                        // the same value appears earlier in the first range; we've already counted occurrences

```

Comment:
Do we have the ability to invoke `_Pred` on two (projected) elements from the first range? I don't see this ability in `indirect_equivalence_relation`. (The classic algorithm required the ranges to have the same value type, but I don't see that requirement for the ranges algorithm.)

If this is not permitted by the concept, then this indicates that we have missing test coverage for differing element types.